### PR TITLE
Ci optimizer

### PR DESCRIPTION
**Pull request by thomas-niehaus**
Optimizes locations of conical intersection between arbitrary neighbouring states. 

- A bit ugly. Had to bring the couplings and exc derivatives down to the level of main. Moving the ground state derivs up to linrespgrad was not possible, because gs gradients are formed after call to casida routines. 
- Parser section could be improved. 
- Getting a strange error in test cyclopentadienyl: Invalid modifier: Debye (haven't touched that part)

### CODE REVIEW

File: src/dftbp/dftbplus/parser.F90 [hunk 1]
```diff
@@ -4866,6 +4866,10 @@ subroutine readExcited(node, geo, ctrl)
       call getChildValue(child, "WriteDensityMatrix", ctrl%lrespini%tWriteDensityMatrix, .false.)
       call getChildValue(child, "WriteXplusY", ctrl%lrespini%tXplusY, default=.false.)
       call getChildValue(child, "StateCouplings", ctrl%lrespini%indNACouplings, default=[0, 0])
+      !!call getChildValue(child, "OptimizeCI", ctrl%lrespini%tCIopt, default=.false.)
+      !!call getChildValue(child, "EnergyShiftCI", ctrl%lrespini%energyShiftCI,  modifier=modifier,&
+      !!    & default=0.0_dp)
+      call convertUnitHsd(char(modifier), energyUnits, child, ctrl%lrespini%energyShiftCI)
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply src/dftbp/dftbplus/parser.F90:4872]: This is the culprit for the Debye error.

**Comment by thomas-niehaus** [comment]
Ben, i tried the test Furan-CIopt-LC with StateCouplings{0 1} and longer optimization. Everything seems to work as expected. What do you observe exactly?

File: src/dftbp/timedep/linresp.F90 [hunk 2]
```diff
@@ -87,7 +87,13 @@ module dftbp_timedep_linresp
 
     !> write X+Y vector sqrt(wij) / sqrt(omega) * F^ia_I
     logical :: tXplusY
+    
+    !> should CI be optimized
+    logical :: tCIopt
```
Comment (bhourahine) [inline src/dftbp/timedep/linresp.F90:92]:
```suggestion
    logical :: isCIopt
```

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+ParserOptions {
+    ParserVersion = 12
+}            
+
+Parallel {
+  # Allow OMP threads explicitely to test for hybrid parallelisation with
+  # MPI-binary. (Check the manual before using this in production runs!)
+  UseOmpThreads = Yes
```
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:48]: Redundant, as this is non-MPI at the moment.

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd [hunk 4]
```diff
@@ -0,0 +1,50 @@
+Geometry = GenFormat {
+    <<< "geo.gen"
+}
+
+Driver = SteepestDescent{MaxSteps=0}
```
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:5]: Deserves a comment
```suggestion
Driver = SteepestDescent{MaxSteps=0} # as we want a fast test
```

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+}
+
+
+Options {
+    WriteAutotestTag = Yes
+}
+
+ParserOptions {
+    ParserVersion = 12
```
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:42]:
```suggestion
    ParserVersion = 13
```

File: test/app/dftb+/timedep/Furan-CIopt/dftb_in.hsd
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt/dftb_in.hsd:1]: Same comments as as the LC version.

**Review by bhourahine — COMMENTED** [review]
- What happens if excitation symmetry is both?
- Does "StateOfInterest" matter for the CI code? Can give input where it is 1 but the state couplings are, e.g. 2 & 3. What state makes sense? (presumably in the example you want to know about 2 and 3).
- Can you suggest approx. magnitude of the values for the energy shift? I.e. just enough to break degeneracy, eV or H?

Please also fix conflicts with main for the changelong and bib file.

File: src/dftbp/dftbplus/initprogram.F90 [hunk 6]
```diff
@@ -474,6 +474,9 @@ module dftbp_dftbplus_initprogram
     !> Is the contribution from an excited state needed for the forces
     logical :: tCasidaForces
 
+    !> Optimization of conical intersections
+    logical :: tCIopt = .false.
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/dftbplus/initprogram.F90:478]: - What happens if excitation symmetry is both? An error is thrown.
- Does "StateOfInterest" matter for the CI code? No. I added a warning if it is set by the user. (This does not catch the user input StateOfInterest=0, however) Also added a clarification on the Total Energies (it is usually the one of StateOfInterest for normal excited state optimizations.)
- Can you suggest approx. magnitude of the values for the energy shift? I.e. just enough to break degeneracy, eV or H? I prefer not to. Too system specific. 
- Changed tCIopt everywhere.

File: src/dftbp/dftbplus/initprogram.F90 [hunk 7]
```diff
@@ -4652,7 +4658,7 @@ subroutine initArrays(this, env, input)
     type(TInputData), intent(in) :: input
 
     logical :: isREKS
-    integer :: nSpinHams, sqrHamSize, iDet
+    integer :: nSpinHams, sqrHamSize, iDet, dLev
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/dftbplus/initprogram.F90:4661]: Gave a comment. It is used only here anyway.

File: doc/dftb+/manual/dftbp.tex [hunk 8]
```diff
@@ -5317,6 +5318,38 @@ \subsection{Casida}
...
+in question is written as entry \is{Energy gap CI} to the output.
+
+\iscb{Bearpark} has one option, which is relevant for the CI between the
+ground and first excited state: 
+  \begin{ptable}
+    \kw{EnergyShift}     & r & & 0.0 & \\
+  \end{ptable}
+  \begin{description}
+  \item[\is{EnergyShift}] Shift of the excited state PES during
+    optimization. Should be brought to zero.
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply doc/dftb+/manual/dftbp.tex:5346]: ok

File: doc/dftb+/manual/dftbp.tex [hunk 9]
```diff
@@ -5317,6 +5318,38 @@ \subsection{Casida}
...
+    \kw{EnergyShift}     & r & & 0.0 & \\
+  \end{ptable}
+  \begin{description}
+  \item[\is{EnergyShift}] Shift of the excited state PES during
+    optimization. Should be brought to zero. 
+  \end{description}
+As discussed by Harabuchi et al.~\cite{Harabuchi2019}, \is{EnergyShift} helps to avoid
+SCC convergence issues close to a ground-to-excited CI. It is
+recommended to first approach the CI with a larger value of
+\is{EnergyShift}, stop the optimization before the loss of convergence,
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply doc/dftb+/manual/dftbp.tex:5351]: ok

File: src/dftbp/timedep/linresp.F90 [hunk 10]
```diff
@@ -202,11 +208,16 @@ subroutine LinResp_init(this, ini, nAtom, nEl, nSpin, onSiteMatrixElements)
...
+        call error("CI optimization: States must be neighbouring.")
+      end if
       this%tNaCoupling = .true.
       this%indNACouplings = ini%indNACouplings
       dLev = ini%indNACouplings(2) - ini%indNACouplings(1)
     endif
-
+    
+    this%tCIopt = ini%tCIopt
+    this%energyShiftCI = ini%energyShiftCI
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/timedep/linresp.F90:220]: Done!

File: src/dftbp/dftbplus/parser.F90 [hunk 11]
```diff
@@ -4893,6 +4893,20 @@ subroutine readExcited(node, geo, ctrl)
...
       end select
 
+      call getChildValue(child, "OptimiserCI", child2, default="goo")
+      call getNodeName(child2, buffer)
+      select case(char(buffer))
+      case ("bearpark")
+        ctrl%lrespini%tCIopt = .true. 
+        call getChildValue(child2, "EnergyShift", ctrl%lrespini%energyShiftCI,  modifier=modifier,&
+          & default=0.0_dp)
+        call convertUnitHsd(char(modifier), energyUnits, child, ctrl%lrespini%energyShiftCI)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/dftbplus/parser.F90:4903]: Done in lresp.F90.

File: src/dftbp/dftbplus/parser.F90 [hunk 12]
```diff
@@ -4893,6 +4893,20 @@ subroutine readExcited(node, geo, ctrl)
         call detailedError(child2, "Invalid diagonaliser method '" // char(buffer) // "'")
       end select
 
+      call getChildValue(child, "OptimiserCI", child2, default="goo")
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/dftbplus/parser.F90:4896]: getChildValue does not seem to work without a specific default value. Goo is just a placeholder. I am open to more elegant solutions.

File: src/dftbp/timedep/linrespgrad.F90 [hunk 13]
```diff
@@ -167,7 +172,7 @@ subroutine LinRespGrad_old(this, iAtomStart, grndEigVecs, grndEigVal, sccCalc, d
     real(dp), allocatable :: xpy(:,:), xmy(:,:), sqrOccIA(:)
     real(dp), allocatable :: xpym(:), xpyn(:), xmyn(:), xmym(:)
     real(dp), allocatable :: t(:,:,:), rhs(:), woo(:,:), wvv(:,:), wov(:)
-    real(dp), allocatable :: eval(:),transitionDipoles(:,:), nacv(:,:,:)
+    real(dp), allocatable :: eval(:),transitionDipoles(:,:)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/timedep/linrespgrad.F90:175]: ok

File: src/dftbp/timedep/linrespgrad.F90 [hunk 14]
```diff
@@ -5451,6 +5460,91 @@ subroutine fixNACVPhase(nacv)
     end do
     
   end subroutine fixNACVPhase
+  
+  !> Implements the CI optimizer of Bearpark et al. Chem. Phys. Lett. 223 269 (1994) with
+  !> modifications introduced by Harabuchi/Hatanaka/Maeda CPL X 2019
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/timedep/linrespgrad.F90:5465]: ok

File: src/dftbp/timedep/linrespgrad.F90 [hunk 15]
```diff
@@ -5451,6 +5460,91 @@ subroutine fixNACVPhase(nacv)
...
+  !> Previous published results [Niehaus JCP 158 054103 (2023), TCA 140 34 (2021)] were obtained
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/timedep/linrespgrad.F90:5466]: ok

File: src/dftbp/timedep/linrespgrad.F90 [hunk 16]
```diff
@@ -5451,6 +5460,91 @@ subroutine fixNACVPhase(nacv)
...
+  !> with a differing version that assumed orthogonal X1 and X2 vectors, which leads to poor convergence
+  subroutine conicalIntersectionOptimizer(derivs, excDerivs, indNACouplings, energyShift, naCouplings, excEnergies)
+
+    !> Ground state gradient (overwritten)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/timedep/linrespgrad.F90:5470]: ok

File: doc/dftb+/manual/manual.bib [hunk 17]
```diff
@@ -1057,3 +1057,24 @@ @article{niehaus2023exact
...
+  year = {1994},
+  volume = {223},
+  pages = {269--274},
+  number = {3},
+  publisher = {Elsevier}
+}
+@ARTICLE{Harabuchi2019,
+  author = {Harabuchi, Yu and Hatanaka, Miho and Maeda, Satoshi},
+  title = {Exploring approximate geometries of minimum energy conical intersections
+	by TDDFT calculations},
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply doc/dftb+/manual/manual.bib:1074]: ok

File: doc/dftb+/manual/manual.bib [hunk 18]
```diff
@@ -1057,3 +1057,24 @@ @article{niehaus2023exact
...
+  journal = {Chemical Physics Letters: X},
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply doc/dftb+/manual/manual.bib:1075]: ok

File: doc/dftb+/manual/manual.bib [hunk 19]
```diff
@@ -1057,3 +1057,24 @@ @article{niehaus2023exact
   year={2023},
   publisher={AIP Publishing LLC}
 }
+@ARTICLE{Bearpark1994,
+  author = {Bearpark, Michael J and Robb, Michael A and Schlegel, H Bernhard},
+  title = {A direct method for the location of the lowest energy point on a
+	potential surface crossing},
+  journal = {Chem. Phys. Lett.},
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply doc/dftb+/manual/manual.bib:1064]: ok

File: src/dftbp/timedep/linresp.F90 [hunk 20]
```diff
@@ -67,7 +67,7 @@ module dftbp_timedep_linresp
     !> atom resolved Hubbard U
     real(dp), allocatable :: HubbardU(:)
 
-    !> atom resolved spin constants
+    !> atom resolved spinconstants
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply src/dftbp/timedep/linresp.F90:70]: ok

File: src/dftbp/timedep/linresp.F90 (hunk 2 above)
Comment (thomas-niehaus) [reply src/dftbp/timedep/linresp.F90:92]: ok

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd (hunk 3 above)
Comment (thomas-niehaus) [reply test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:48]: ok

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd (hunk 4 above)
Comment (thomas-niehaus) [reply test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:5]: ok

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd (hunk 5 above)
Comment (thomas-niehaus) [reply test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:42]: ok

File: src/dftbp/dftbplus/parser.F90 (hunk 12 above)
[the remark this answers is not in the document]
Comment (bhourahine) [reply src/dftbp/dftbplus/parser.F90:4909]: What should work is (you'll need type(fnode) child3 to be defined):

call getChildValue(child, "OptimiserCI", child2, "", child=child3, allowEmptyValue=.true.)
if (associated(child2)) then
  call getNodeName(child2, buffer)
.
.
.
end if

**Comment by bhourahine** [comment]
Looks like there is now a problem with timedep/Furan-NACV :

ERROR!
-> CI optimization: States must be neighbouring.

File: src/dftbp/timedep/linrespgrad.F90 [hunk 21]
```diff
@@ -649,61 +663,61 @@ subroutine LinRespGrad_old(this, iAtomStart, grndEigVecs, grndEigVal, sccCalc, d
...
-              & tRangeSep, rangeSep, excgrad)
-        end if
+          if (this%writeMulliken) then
+            !> for now, only total Mulliken charges
+            call writeExcMulliken(sym, iLev, dq(:,1), sum(dqex,dim=2), coord0)
+          end if
 
-      end do
+          if (tForces) then
+            iSav = iLev - nStartLev + 1
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:708]:
```suggestion
            iSav = iLev - nStartLev + 1
```

File: src/dftbp/timedep/linrespgrad.F90 [hunk 22]
```diff
@@ -799,19 +814,23 @@ subroutine LinRespGrad_old(this, iAtomStart, grndEigVecs, grndEigVal, sccCalc, d
...
-
-      if (nstat == 0) then
-        omega = 0.0_dp
+            
       end if
 
     end if
 
+    !> Omega has possibly been overwritten for CI optimization or NA couplings, but should always
+    !! refer to nstat
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:827]:
```suggestion
    !! refer to nstat
```

File: src/dftbp/timedep/linrespgrad.F90 [hunk 23]
```diff
@@ -799,19 +814,23 @@ subroutine LinRespGrad_old(this, iAtomStart, grndEigVecs, grndEigVal, sccCalc, d
...
 
     end if
 
+    !> Omega has possibly been overwritten for CI optimization or NA couplings, but should always
+    !! refer to nstat 
+    if (nstat == 0) then
+      omega = 0.0_dp
+    else
+      omega = sqrt(eval(nstat))
+    end if
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:832]:
```suggestion
    end if
```

File: src/dftbp/timedep/linrespgrad.F90 [hunk 24]
```diff
@@ -5456,6 +5475,91 @@ subroutine fixNACVPhase(nacv)
     end do
     
   end subroutine fixNACVPhase
+
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:5478]:
```suggestion

```

File: src/dftbp/timedep/linrespgrad.F90 [hunk 25]
```diff
@@ -5456,6 +5475,91 @@ subroutine fixNACVPhase(nacv)
...
+    dpX2 = dot_product(X2, X2)
+    dp12 = dot_product(X1, X2)
+    prj1 = dot_product(dE2, X1)
+    prj2 = dot_product(dE2, X2)
+
+    alpa = (prj2*dp12 - prj1*dpX2)/(dpX1*dpX2 - dp12*dp12)
+    beta = (prj2*dpX1 - prj1*dp12)/(dp12*dp12 - dpX1*dpX2)
+
+    ! Eq. 5 in Bearpark et al.
+    gpf(:) = dE2(:) + alpa * X1(:) + beta * X2(:)
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:5546]:
```suggestion
    gpf(:) = dE2(:) + alpa * X1(:) + beta * X2(:)
```

File: src/dftbp/timedep/linrespgrad.F90 [hunk 26]
```diff
@@ -5456,6 +5475,91 @@ subroutine fixNACVPhase(nacv)
...
+    allocate(X2(3 * nAtoms))
+    allocate(dE2(3 * nAtoms))
+    allocate(gpf(3 * nAtoms))
+
+    if (indNACouplings(1) == 0) then
+      X1 = reshape(excDerivs(:,:,nexcGrad), (/ 3 * nAtoms /))
+    else
+      X1 = reshape(excDerivs(:,:,nexcGrad)-excDerivs(:,:,1), (/ 3 * nAtoms /))
+    end if
+    ! Last entry of array holds coupling between states indNACouplings(1/2)
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:5529]:
```suggestion
    ! Last entry of array holds coupling between states indNACouplings(1/2)
```

File: src/dftbp/timedep/linrespgrad.F90 [hunk 27]
```diff
@@ -5456,6 +5475,91 @@ subroutine fixNACVPhase(nacv)
...
+
+    ! Eq. 5 in Bearpark et al.
+    gpf(:) = dE2(:) + alpa * X1(:) + beta * X2(:) 
+    normGP = norm2(gpf)
+
+    ! Eq. 4 in Bearpark et al. with modifications by Harabuchi
+    ! Yields approximate CI without running into SCF problems too early
+    ! Shift should be brought to zero
+    if (indNACouplings(1) == 0) then
+      deltaE = excEnergies(indNACouplings(2))
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:5553]:
```suggestion
      deltaE = excEnergies(indNACouplings(2))
```

File: src/dftbp/timedep/linresptypes.F90 [hunk 28]
```diff
@@ -80,6 +80,12 @@ module dftbp_timedep_linresptypes
     !> Should non-adiabatic couplings be computed?
     logical :: tNaCoupling = .false.
 
+    !> should CI be optimized
+    logical :: isCIopt
+
+    !> Energy shift used in CI optimizer
+    real(dp) :: energyShiftCI
+
```
Comment (bhourahine) [inline src/dftbp/timedep/linresptypes.F90:88]:
```suggestion

```

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+    SlaterKosterFiles = Type2FileNames {
+        Prefix = "slakos/origin/ob2-1-1/shift/"
+        Separator = "-"
+        Suffix = ".skf"
+    }
+    RangeSeparated = LC {
+      Screening = Thresholded {
+        Threshold = 1.0e-16
+      }
+    }
```
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:23]:
```suggestion
    }
```

File: test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+
+Options {
+    WriteAutotestTag = Yes
+}
+
+ParserOptions {
+    ParserVersion = 13
+}
```
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt-LC/dftb_in.hsd:42]:
```suggestion
}
```

File: test/app/dftb+/timedep/Furan-CIopt/dftb_in.hsd [hunk 31]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+
+Options {
+    WriteAutotestTag = Yes
+}
+
+ParserOptions {
+    ParserVersion = 13
+}
```
Comment (bhourahine) [inline test/app/dftb+/timedep/Furan-CIopt/dftb_in.hsd:39]:
```suggestion
}
```

**Review by bhourahine — COMMENTED** [review]
Bunch of trivisl whitespace (should be able to approve all of them on github web interface).

File: src/dftbp/timedep/linrespgrad.F90 [hunk 32]
```diff
@@ -799,17 +813,21 @@ subroutine LinRespGrad_old(this, iAtomStart, grndEigVecs, grndEigVal, sccCalc, d
...
         end do
 
-        !> Convention to determine arbitrary phase 
-        call fixNACVPhase(nacv) 
+        !> Convention to determine arbitrary phase
+        call fixNACVPhase(nacv)
 
         call writeNACV(this%indNACouplings(1), this%indNACouplings(2), fdTagged, taggedWriter, nacv)
-
+
```
Comment (bhourahine) [inline src/dftbp/timedep/linrespgrad.F90:820]:
```suggestion

```

**Review by bhourahine — APPROVED** [review]
OK, ready to go.

**Pull request merged**